torch/onnx/_internal/fx/passes/modularization.py

```diff
@@ -0,0 +1,648 @@
+        self, module_name: str, module_class: Optional[type], raw_meta: Tuple[Any, Any]
+    ):
+        self._module_name = module_name
+        self._module_class = module_class
+        self._raw_meta = raw_meta
+
+    @property
+    def module_display_name(self) -> str:
+        """The display name of the module.
+
+        E.g. `h_1_mlp_c_proj`.
+        """
+        # E.g., from 'L__self___h_1_mlp_c_proj' to 'h_1_mlp_c_proj'.
+        name = self.module_name
+        if name.startswith("L__self___"):
+            name = name[len("L__self___") :]
+        return name
+
+    @property
+    def qualified_module_class_name(self) -> str:
+        """Qualified name of the module class.
+
+        E.g. `torch_nn_module_sparse_Embedding`.
+        """
+        if self._module_class is None:
+            return ""
+        return (
+            self._module_class.__module__ + "_" + self._module_class.__name__
+        ).replace(".", "_")
+
+    @property
+    def module_class_name(self) -> str:
+        """Name of the module class.
+
+        E.g. `Embedding`.
+        """
+        if self._module_class is None:
+            return ""
+        return self._module_class.__name__
+
+    @property
+    def module_name(self) -> str:
+        """Name of the module.
+
+        E.g. `L__self___h_1_mlp_c_proj`.
+        """
+        return self._module_name
+
+    @property
+    def raw_meta(self) -> Tuple[Any, Any]:
+        """Returns the raw module meta data.
+
+        I.e. (module_name, node.meta['nn_module_stack'][module_name]).
+        """
+        return self._raw_meta
+
+    def __eq__(self, __value: object) -> bool:
+        if not isinstance(__value, _ModuleMeta):
+            return False
+        return (
+            self._module_name == __value._module_name
+            and self._module_class == __value._module_class
+        )
+
+    def __hash__(self) -> int:
+        return hash((self._module_name, self._module_class))
+
+    def __repr__(self) -> str:
+        return f"ModuleMeta(name={self._module_name}, class={self._module_class})"
+
+    @classmethod
+    def root(cls) -> _ModuleMeta:
+        """Create an empty module meta representing root module."""
+        return _ModuleMeta("", None, ("", None))
+
+    @classmethod
+    def from_fx_tracer_produced_raw_meta(
+        cls, raw_meta: Tuple[str, Type]
+    ) -> _ModuleMeta:
+        """Create a module meta from raw meta produced by FX symbolic tracer."""
+        module_name, module_class = raw_meta
+        return _ModuleMeta(module_name, module_class, raw_meta)
+
+    @classmethod
+    def from_dynamo_produced_raw_meta(
+        cls, raw_meta: Tuple[str, Tuple[str, Type]]
+    ) -> _ModuleMeta:
+        """Create a module meta from raw meta produced by FX dynamo tracer."""
+        module_name, (qualified_name, module_class) = raw_meta
+        return _ModuleMeta(module_name, module_class, raw_meta)
+
+
+class _ModuleStackMeta:
+    """Meta information about the module stack.
```

Comment:
What is `module stack`? I guess it means the `model hierarchy` like
```
GPT(nn.Module)
  SelfAttentionLayer(nn.Module)
     Gemm
```
? But it could also mean the `call stack` invoked to reach a module's forward method. Could you make it clear? Thanks.

Reply:
name inherited from `nn_module_stack` meta. Updated docstring to make it more obvious.
And you are right, it is call stack.

Prefers current name, since it has matching raw meta.